ffi/src/handle.rs

```diff
@@ -1,235 +1,527 @@
+//! advised to maintain "unique pointer" semantics for mutable handles.
+//!
+//! NOTE: While shared handles could conceptually impl [`Clone`], cloning would require unsafe code
+//! and so we can't actually implement the trait. Use [`CloneHandle::clone_handle`] instead.
+use std::ptr::NonNull;
 use std::sync::Arc;
 
-/// Helper trait that simplifies passing an instance of a Sized type across the FFI boundary as a
-/// leaked thin pointer. Does not include any reference-counting capability, so engine is
-/// responsible to do whatever reference counting the engine may need. Engine is also responsible
-/// not to drop an in-use handle, so kernel code can safely dereference the pointer.
-pub trait BoxHandle: Sized + Send + Sync {
-    fn into_handle(self) -> *mut Self {
-        Box::into_raw(Box::new(self))
-    }
+/// Provides mutable access to the underlying object of a mutable [`Handle`] .
+/// Similar to [`AsMut`], but unsafe and hard-wired to the handle's target type.
+pub trait HandleAsMut {
+    /// The true type of the underlying object.
+    type Target: ?Sized;
+
+    /// Obtains a mutable reference to the handle's underlying object.
+    ///
     /// # Safety
     ///
-    /// The `handle` was returned by a call to `into_handle`, has not already been passed to
-    /// `drop_handle`, and has no other live references.
-    unsafe fn drop_handle(handle: *mut Self) {
-        let _ = unsafe { Box::from_raw(handle) };
-    }
+    /// Caller asserts that this handle obeys [mutable reference
+    /// semantics](https://doc.rust-lang.org/book/ch04-02-references-and-borrowing.html#the-rules-of-references),
+    /// namely:
+    ///
+    /// * The handle is [valid][Handle#Validity].
+    ///
+    /// * No other references (neither shared nor mutable) can overlap with the returned reference.
+    unsafe fn as_mut(&mut self) -> &mut Self::Target;
 }
 
-mod unconstructable {
-    /// A struct that cannot be instantiated by any rust code because this module exposes no public
-    /// constructor for it. Intentionally _NOT_ a zero-sized type, to avoid weirdness in C/C++ land.
-    pub struct Unconstructable {
-        _private: usize,
-    }
-}
+/// Clones a a shared [`Handle`]. Similar to [`Clone`], but unsafe. Also allows to clone the handle
+/// as an [`Arc`] for easy access to the handle's underlying object.
+pub trait CloneHandle {
+    /// The true type of the underlying object.
+    type Target: ?Sized;
 
-// Make it easy to use (the module was only used to enforce member privacy).
-pub type Unconstructable = unconstructable::Unconstructable;
+    /// Clones this handle, increasing the underlying object's refcount by one. The new handle is
+    /// independent of the source handle, and should be dropped when it is no longer needed.
+    ///
+    /// # Safety
+    ///
+    /// Caller asserts that the handle is [valid][Handle#Validity].
+    unsafe fn clone_handle(&self) -> Self;
```

Comment:
This isn't used. Can we call it `_clone_handle` or `#[allow(dead_code)]` it

Reply:
I'm a bit confused actually -- this is a `pub trait` whose methods therefore are all public. 
Why would it even trigger dead code warnings in the first place?

Reply:
Sorry, that'll happen if you make the other change I suggested where handle is `pub(crate) mod handle;`